### DISCUSSION TITLE
Overriding Input Schemas to allow for non-object inputs

### DISCUSSION THREAD

**Discussion by TheWisestOne** [RobinTail/express-zod-api#897] [Q&A] [answered]
Hi,

Working on migrating an existing API to express-zod-api for better documentation and type enforcement, but running into an issue with an existing endpoint.

I have a `PUT` request that takes as input an `Array` of objects. This array is sent *directly* on the body of the request, not wrapped in a JSON object. I understand this is often not considered a best practice, and we have plans to upgrade this api eventually to follow a more standard convention, but for the purpose of not making breaking changes, I would like to support this API as-is.

For a similar thing on the response side, I know we can override the `ResultHandler`, but I can't figure out how to do the same thing on the input side. I have thought of a few things like adding middlewares *before* the actual API handler to massage the input into the correct format but can't seem to get to anything that works. Is this possible to do, or do we just have to bite the bullet and make a breaking change here?

**Comment by RobinTail** [comment, marked as answer]
Hello @TheWisestOne ,
Middlewares provide you with direct access to request object. You can check it with Array.isArray() and return as a property inside the options, accessible by endpoints. 
All that is beyond the proper types support.
But it might work this way.

**Reply by TheWisestOne** [reply]
Thanks for the response! I had actually already tried something similar to that, but the issue is that even though I can access the data, it never even gets to execute the handler because the input validation fails, since the request body doesn't adhere to IOSchema. The error happens in endpoint.ts

I also tried adding a raw express Middleware at the beginning, which essentially does:

`request.body = { array: request.body }`

Hoping to update the body before it got to that part of the execution...however, it seems in endpoint.ts, getInput runs before any of the Middleware. Let me know if there's something I missed!

**Reply by RobinTail** [reply]
Ah... yeah, that's correct. Middleware has the input validation too.
I need to think a bit more on this

**Reply by TheWisestOne** [reply]
Hi @RobinTail, I was actually able to get this to work with something close to your original suggestion.

```
export const handleArrayBodyApi = endpointsFactory
    .addMiddleware(requestProviderMiddleware)
    .build({
        method: 'put',
        tag: '',
        description: 'update using array',
        input: z.object({
            id: z.string().uuid(),
        }),
        output: z.object({
            boolean: z.boolean(),
        }),
        async handler({ input: { id }, options: { request } }) {
            const completed = await useRequestBodyAsArray(
                id,
                request.body,
            )
            return { boolean: completed }
        },
    })
```

I'm not sure what was broken the first time I tried it, but if I simply never try to enforce any type constraints on`request.body`, and have a middleware that just returns `request` as-is, everything seems to work correctly. Thanks for your help!

**Reply by RobinTail** [reply]
Cool! Good to know, @TheWisestOne .
Thank you for the code sample
